bigframes/ml/compose.py

```diff
@@ -46,6 +48,114 @@
 )
 
 
+CUSTOM_TRANSFORMER_SQL_RX = re.compile(
+    "^(?P<sql>.*)/[*]CT.(?P<id>[A-Z]+[A-Z0-9]*)[(](?P<config>[^*]*)[)][*]/$",
+    re.IGNORECASE,
+)
+
+
+class CustomTransformer(base.BaseTransformer):
+    _CTID = None
+    _custom_transformer_classes = {}
+
+    @classmethod
+    def register(cls, transformer_cls: Type[base.BaseTransformer]):
+        assert transformer_cls._CTID
+        cls._custom_transformer_classes[transformer_cls._CTID] = transformer_cls
+
+    @classmethod
+    def find_matching_transformer(
+        cls, transform_sql: str
+    ) -> Optional[Type[base.BaseTransformer]]:
+        for transform_cls in cls._custom_transformer_classes.values():
+            if transform_cls.understands(transform_sql):
+                return transform_cls
+        return None
+
+    @classmethod
+    def understands(cls, transform_sql: str) -> bool:
+        """
+        may be overwritten to have a more advanced matching, possibly without comments in SQL
+        """
+        m = CUSTOM_TRANSFORMER_SQL_RX.match(transform_sql)
+        if m and m.group("id").strip() == cls._CTID:
+            return True
+        return False
+
+    def __init__(self):
+        super().__init__()
+
+    def _compile_to_sql(
+        self, X: bpd.DataFrame, columns: Optional[Iterable[str]] = None
+    ) -> List[str]:
+        if columns is None:
+            columns = X.columns
+        return [
+            f"{self.custom_compile_to_sql(X, column)} {self._get_sql_comment(column)} AS {self.get_target_column_name(column)}"
+            for column in columns
+        ]
+
+    def get_target_column_name(self, column: str) -> str:
+        return f"{self._CTID.lower()}_{column}"
+
+    @classmethod
+    @abc.abstractclassmethod
+    def custom_compile_to_sql(cls, X: bpd.DataFrame, column: str) -> str:
+        pass
+
+    def get_persistent_config(self, column: str) -> Optional[Union[Dict, List]]:
+        """
+        return structure to be persisted in the comment of the sql
+        """
+        return None
+
+    def _get_pc_as_args(self, column: str) -> str:
+        config = self.get_persistent_config(column)
+        if not config:
+            return ""
+        return json.dumps(config)
+
+    def _get_sql_comment(self, column: str) -> str:
+        args = self._get_pc_as_args(column)
+        return f"/*CT.{self._CTID}({args})*/"
+
+    @classmethod
+    def _parse_from_sql(cls, transform_sql: str) -> Tuple[base.BaseTransformer, str]:
+        m = CUSTOM_TRANSFORMER_SQL_RX.match(transform_sql)
+        if m and m.group("id").strip() != cls._CTID:
+            raise ValueError("understand() does not match _parse_from_sql!")
+        args = m.group("config").strip()
+        if args != "":
+            config = json.loads(args)
+        else:
+            config = None
+        sql = m.group("sql").strip()
+        return cls.custom_parse_from_sql(config, sql)
+
+    @classmethod
+    @abc.abstractclassmethod
+    def custom_parse_from_sql(
```

Comment:
How important is this feature to you? Do you expect folks will need to inspect the learned parameters even after loading the trained transform from BigQuery? Without this feature, folks would still be able to inspect parameters after calling `fit()`.

Reply:
I could think about a VertexAI-Pipeline (Kubeflow), which creates a transform-only model in the first component and in following steps it tests different estimators and finally creates a ml.Pipeline which contains the transformation together with the best estimator together in one Model. So, the final model contains the initially created tranformation and the prediction together.

Each VertexAI component is executed in a seperate process, so the transformation created using fit() is not available in the succeeding steps. It has to be loaded.

But we could simplify things and instead of parsing sql we could only use the persisted config. 
Parsing the sql was neccessary to retrieve the original column label. 
But this could also be added to the persisted config `get_persistent_config()`.

Reply:
I see you use `TRANSFORM_ONLY` models in the description of this PR. I'll try to use such a model with `read_gbq_model()`. If so, I'm wondering if it's necessary to save the parameters in comments.

Reply:
I see we have a limitation currently in BigQuery DataFrames. I created a `TRANSFORM_ONLY` model

```
CREATE OR REPLACE MODEL `swast-scratch`.bigframes_pr955.species_length
TRANSFORM(CASE WHEN species IS NULL THEN -1 ELSE LENGTH(species) END AS len1_species)
  OPTIONS (
    model_type = 'transform_only')
AS (
  SELECT *
  FROM `bigquery-public-data.ml_datasets.penguins`
);
```

But this "model" failed to load with `len1_species = bpd.read_gbq_model("swast-scratch.bigframes_pr955.species_length")`. It's failing on this assertion: https://github.com/googleapis/python-bigquery-dataframes/blob/569a7ad5a2fe72a8d3deb1304eb7180b176d8830/bigframes/ml/compose.py#L168

I will continue to investigate this. IMO we should allow for this kind of TRANSFORM_ONLY model to be used from bigframes.

Reply:
Yes, in my sample code I saved the ml.ColumnTransformer and not a ml.Pipeline.
When reading the model with read_gbq_model() the ColumnTransformer is returned again, execpt for the case, that all columns are transformed in the same way, then the Transformer itself is "merged" together and the ML.xxx transformer is returned. To simplify the CustomTransformer code I disabled merging if a CustomTransformer is contained.

Putting the parameters into the comment inside the sql of each column is a workaround.
Maybe we can persist this info at another place, e.g. the Description of the model.
Then each CustomTransformer returns the plain SQL in compile_to_sql() and the ColumnTransformer writes the list of params (retrieved with get_persistent_config()) for each CustomTransformer together with the CustomTransformer-ID and the column name into the model description. 

I am fine, if ColumnTransformer is not supporting the edge case of an empty transformation.